src/persist/src/indexed/mod.rs

```diff
@@ -1291,6 +1400,150 @@ mod tests {
+        let blob = registry.blob_no_reentrance()?;
+        let blob = UnreliableBlob::from_handle(blob, unreliable.clone());
+        let blob = BlobCache::new(build_info::DUMMY_BUILD_INFO, metrics.clone(), blob);
+        let maintainer = Maintainer::new(blob.clone(), Arc::new(Runtime::new()?));
+        let mut i = Indexed::new(log, blob, metrics)?;
+
+        let updates: Vec<((Vec<u8>, Vec<u8>), u64, isize)> = vec![
+            (("1".into(), "".into()), 1, 1),
+            (("2".into(), "".into()), 2, 1),
+            (("3".into(), "".into()), 3, 1),
+            (("4".into(), "".into()), 4, 1),
+            (("5".into(), "".into()), 5, 1),
+        ];
+
+        let reqs = i.step()?;
+        // No trace compaction requests when no streams are registered.
+        assert_eq!(reqs, vec![]);
+
+        let id = block_on(|res| i.register("0", "()", "()", res))?;
+
+        let reqs = i.step()?;
+        // No maintenace requests when the registered stream has no data
+        // in the trace.
+        assert_eq!(reqs, vec![]);
+
+        // Move data into trace.
+        block_on_drain(&mut i, |i, handle| {
+            i.write(
+                vec![(id, updates.iter().collect::<ColumnarRecords>())],
+                handle,
+            )
+        })?;
+
+        for t in 2..7 {
+            block_on_drain(&mut i, |i, handle| i.seal(vec![id], t, handle))?;
+            let reqs = i.step()?;
+            // There will not be any compaction requests because the compaction
+            // frontier has not advanced.
+            assert_eq!(reqs, vec![]);
+        }
+
+        // Advance the compaction frontier to a time that has been sealed.
+        block_on_drain(&mut i, |i, handle| {
+            i.allow_compaction(vec![(id, Antichain::from_elem(5))], handle)
+        })?;
+
+        let reqs = i.step()?;
+        assert_eq!(reqs.len(), 1);
+        let request = reqs[0].clone();
+
+        // Once a request is in flight, don't send requests for the same stream.
+        let reqs = i.step()?;
+        assert_eq!(reqs, vec![]);
+
+        // Handle receiving an error MaintenanceRes.
+        let error_response =
+            MaintenanceRes::CompactTrace((id, Err(Error::from("test compaction error"))));
+        i.handle_maintenance_res(error_response.clone());
+        assert_eq!(i.drain_pending(), Ok(()));
+
+        // After receiving an error, retry the original request.
+        let reqs = i.step()?;
+        assert_eq!(reqs.len(), 1);
+        assert_eq!(reqs[0], request);
+
+        // Handle receiving a valid MaintenanceRes.
+        let response = futures_executor::block_on(request.clone().run_async(maintainer.clone()));
+
+        i.handle_maintenance_res(response.clone());
+        assert_eq!(i.drain_pending(), Ok(()));
+
+        // After receiving a valid response, issue a new compaction request.
+        let reqs = i.step()?;
+        assert_eq!(reqs.len(), 1);
+        let request2 = reqs[0].clone();
```

Comment:
I'm a little torn because on one hand, the way this test handles checking requests is lazy and especially asserting on length == 1 does not give good error messages on failure. On the other hand, we already have test coverage in arrangement and background that checks to make sure the requests and responses are handled correctly -- these tests are much more about the at-most-one req in flight bookkeeping and making sure retries happen correctly so they don't care about the contents of the request/response as much.

wdyt? I'm happy to change the asserts to something that more strongly checks the contents of the request/response